gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/main/java/io/gravitee/gateway/jupiter/debug/policy/steps/PolicyStep.java

```diff
@@ -0,0 +1,218 @@
+            .andThen(computeDiff(source, this.inputState, AttributeHelper.filterAndSerializeAttributes(attributes)))
+            .doOnSuccess(
+                computedDiffMap -> {
+                    this.diffMap = computedDiffMap;
+                    this.inputState = null;
+                    this.status = this.status == null ? DebugStepStatus.COMPLETED : this.status;
+                }
+            )
+            .ignoreElement();
+    }
+
+    protected abstract Single<Map<String, Object>> computeDiff(
+        final T source,
+        final PolicyStepState inputState,
+        final Map<String, Serializable> outputAttributes
+    );
+
+    public Map<String, Object> getDiff() {
+        if (diffMap == null) {
+            diffMap = new HashMap<>();
+        }
+        return diffMap;
+    }
+
+    protected PolicyStepState getInputState() {
+        return inputState;
+    }
+
+    protected void start() {
+        if (!stopwatch.isRunning()) {
+            this.stopwatch.start();
+        }
+    }
+
+    protected void stop() {
+        if (stopwatch.isRunning()) {
+            this.stopwatch.stop();
+        }
+    }
+
+    public Completable error(Throwable ex) {
+        return Completable.fromRunnable(
+            () -> {
+                stop();
+                this.status = DebugStepStatus.ERROR;
+                this.error = new DebugStepError();
+                this.error.setMessage(ex.getMessage());
+            }
+        );
+    }
+
+    public Completable error(ExecutionFailure executionFailure) {
+        return Completable.fromRunnable(
+            () -> {
+                stop();
+                this.status = DebugStepStatus.ERROR;
+                this.error = new DebugStepError();
+                this.error.setMessage(executionFailure.message());
+                this.error.setKey(executionFailure.key());
+                this.error.setStatus(executionFailure.statusCode());
+                this.error.setContentType(executionFailure.contentType());
+            }
+        );
+    }
+
+    public void onConditionFilter(String condition, boolean isConditionTruthy) {
+        this.condition = condition;
+        if (!isConditionTruthy) {
+            this.status = DebugStepStatus.SKIPPED;
+        }
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    public String getPolicyId() {
+        return policyId;
+    }
+
+    public ExecutionPhase getExecutionPhase() {
+        return executionPhase;
+    }
+
+    public String getFlowPhase() {
+        return flowPhase;
+    }
+
+    public Duration elapsedTime() {
+        return this.stopwatch.elapsed();
+    }
+
+    public DebugStepStatus getStatus() {
+        return status;
+    }
+
+    public String getCondition() {
+        return condition;
+    }
+
+    public DebugStepError getError() {
+        return error;
+    }
+
+    public boolean isEnded() {
+        return ended;
+    }
+
+    public void ended() {
```

Comment:
I would have chosen a verb to set the variable, instead of `ended` that I would have use as `isEnded`
```suggestion
    public void end() {
```